src/daft-minhash/src/minhash.rs

```diff
@@ -0,0 +1,215 @@
+        for _ in 0..SIMD_LANES {
+            *o = fast_simd_rem(h.mul(*a).add(*b)).simd_min(*o);
+            h = h.rotate_elements_left::<1>();
+        }
+    }
+}
+
+#[inline(always)]
+fn rem_min(hh: &[u64], aa: &[u64], bb: &[u64], out: &mut [u64]) {
+    for h in hh {
+        for ((a, b), o) in aa.iter().zip(bb.iter()).zip(out.iter_mut()) {
+            for _ in 0..SIMD_LANES {
+                *o = h
+                    .mul(*a)
+                    .add(*b)
+                    .rem(MERSENNE_PRIME)
+                    .bitand(MAX_HASH)
+                    .min(*o);
+            }
+        }
+    }
+}
+
+pub fn load_simd(v: &[u64]) -> Vec<S> {
+    let num_hashes = v.len();
+    let num_hashes_aligned = (num_hashes + SIMD_LANES - 1) / SIMD_LANES * SIMD_LANES;
+    let num_simd = num_hashes_aligned / SIMD_LANES;
+
+    let mut v_simd: Vec<S> = Vec::with_capacity(num_simd);
+    for i in 0..num_simd {
+        v_simd.push(S::load_or_default(&v[(SIMD_LANES * i)..]));
+    }
+    v_simd
+}
+
+pub fn minhash(
+    s: &str,
+    perm: (&[u64], &[u64]),
+    perm_simd: (&[S], &[S]),
+    ngram_size: usize,
+    seed: u32,
+) -> DaftResult<Vec<u32>> {
+    let (perm_a, perm_b) = perm;
+    let (perm_a_simd, perm_b_simd) = perm_simd;
+    let num_hashes = perm_a.len();
+    let num_simd = (num_hashes + SIMD_LANES - 1) / SIMD_LANES;
+
+    let mut out: Vec<S> = vec![MAX_HASH_SIMD; num_simd];
+
+    let spaces: Vec<usize> = s.match_indices(' ').map(|(i, _)| i).collect();
+    let ngram_count = if spaces.len() < ngram_size {
+        1
+    } else {
+        spaces.len() - ngram_size + 2
+    };
+    let mut hashes: Vec<u64> = Vec::with_capacity(SIMD_LANES);
+    let s_bytes = s.as_bytes();
+    if spaces.len() < ngram_size {
+        // hash whole string at once
+        hashes.push(murmurhash3_x86_32(s_bytes, seed) as u64);
+    } else {
+        for i in 0..ngram_count {
+            // looking at the substring that starts BEFORE the current space
+            // surely no off by one errors
+            let start_ind = if i == 0 { 0 } else { spaces[i - 1] + 1 };
+            let end_ind = if i == ngram_count - 1 {
+                s.len()
+            } else {
+                spaces[i + ngram_size - 1]
+            };
+            hashes.push(murmurhash3_x86_32(&s_bytes[start_ind..end_ind], seed) as u64);
+            if hashes.len() >= SIMD_LANES {
+                let hashes_simd = S::from_slice(&hashes);
+                simd_min(hashes_simd, perm_a_simd, perm_b_simd, &mut out);
+                hashes.clear();
+            }
+        }
+    }
+
+    let mut rem_out: Vec<u64> = out
+        .iter()
+        .flat_map(|x| x.as_array())
+        .take(num_hashes)
+        .copied()
+        .collect();
+    if !hashes.is_empty() {
+        rem_min(&hashes, perm_a, perm_b, &mut rem_out);
+    }
+    Ok(rem_out.into_iter().map(|x| x as u32).collect())
+}
+
+// cargo bench --package daft-minhash
+#[cfg(test)]
+mod tests {
```

Comment:
Are these tests already automatically run in the CI or do I need to add them somehow?